packages/wallets/exodus/src/adapter.ts

```diff
@@ -0,0 +1,268 @@
+
+    private _connecting: boolean;
+    private _wallet: ExodusWallet | null;
+    private _publicKey: PublicKey | null;
+    private _readyState: WalletReadyState =
+        typeof window === 'undefined' || typeof document === 'undefined'
+            ? WalletReadyState.Unsupported
+            : WalletReadyState.NotDetected;
+
+    constructor(config: ExodusWalletAdapterConfig = {}) {
+        super();
+        this._connecting = false;
+        this._wallet = null;
+        this._publicKey = null;
+
+        if (this._readyState !== WalletReadyState.Unsupported) {
+            scopePollingDetectionStrategy(() => {
+                if (window.solana?.isExodus) {
+                    this._readyState = WalletReadyState.Installed;
+                    this.emit('readyStateChange', this._readyState);
+                    return true;
+                }
+                return false;
+            });
+        }
+    }
+
+    get publicKey(): PublicKey | null {
+        return this._publicKey;
+    }
+
+    get connecting(): boolean {
+        return this._connecting;
+    }
+
+    get connected(): boolean {
+        return !!this._wallet?.isConnected;
+    }
+
+    get readyState(): WalletReadyState {
+        return this._readyState;
+    }
+
+    async connect(): Promise<void> {
+        try {
+            if (this.connected || this.connecting) return;
+            if (this._readyState !== WalletReadyState.Installed) throw new WalletNotReadyError();
+
+            this._connecting = true;
+
+            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+            const wallet = window!.solana!;
+
+            if (!wallet.isConnected) {
+                // HACK: Exodus doesn't reject or emit an event if the popup is closed
+                const handleDisconnect = wallet._handleDisconnect;
```

Comment:
Hmm this code looks like it was copied from the Phantom adapter. The point of this `HACK` comment is to show a hack around a broken behavior in Phantom. You probably don't want to copy it, or the behavior. :) Any code that's in here should pertain to your actual wallet extension/app (which hopefully isn't broken).

Reply:
That is correct. As we made our Solana Provider OOTB compatible with Phantom's when it comes to signatures and return values, we copied the Phantom adapter verbatim. We've made improvements though, like throwing a `UserRejectedRequestError` with `code=4001` if the popup is closed in this case. Cleaning up.